### PR TITLE
Support Highfreq Backtest with the Model/Rule/RL Strategy

### PR DESCRIPTION
**Pull request by bxdd**
## Description

Support Highfreq Backtest with the Model/Rule/RL Strategy

## How Has This Been Tested?
- [ ] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [x] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [ ] Fix bugs
- [x] Add new feature
- [ ] Update documentation

### CODE REVIEW

File: qlib/contrib/backtest/__init__.py [hunk 1]
```diff
@@ -2,12 +2,12 @@
 # Licensed under the MIT License.
 
 from .order import Order
```
[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/contrib/backtest/__init__.py:4]: You can move them later

File: qlib/contrib/backtest/__init__.py [hunk 2]
```diff
@@ -149,176 +71,64 @@ def get_exchange(
...
+            if "sub_strategy" in env_kwargs:
+                env_kwargs["sub_strategy"] = init_instance_by_config(env_kwargs["sub_strategy"])
+            env_config["kwargs"] = env_kwargs
+        return init_instance_by_config(env_config)
+    else:
+        return env
+
+
+def setup_exchange(root_instance, trade_exchange=None, force=False):
+    if "trade_exchange" in inspect.getfullargspec(root_instance.__class__).args:
```
[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/contrib/backtest/__init__.py:109]: default exchange

File: qlib/contrib/backtest/__init__.py [hunk 3]
```diff
@@ -149,176 +71,64 @@ def get_exchange(
...
-
-        .. note:: This will be faster with offline qlib.
-
-    - **executor related arguments**
+        return exchange
+    else:
+        return init_instance_by_config(exchange, accept_types=Exchange)
+
+
+def init_env_instance_by_config(env):
```
[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/contrib/backtest/__init__.py:93]: image

**Comment by bxdd** [comment]
I add a notebook file `workflow.ipynb` to show the multi-level reports, its size is 5k. I temporarily modified .gitignore to submit it successfully. And this file has the same purpose as `workflow_by_code.ipynb`. Is this file necessary? If not, I will reset the related commit. @you-n-g

File: qlib/strategy/base.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+from ..model.base import BaseModel
+from ..data.dataset import DatasetH
+from ..data.dataset.utils import convert_index_format
+from ..contrib.backtest.order import Order
+from ..contrib.backtest.executor import BaseTradeCalendar
+from ..rl.interpreter import ActionInterpreter, StateInterpreter
+
+
+class BaseStrategy(BaseTradeCalendar):
```
Comment (ultmaster) [inline qlib/strategy/base.py:16]: Strategy should not stateless.

File: qlib/strategy/base.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        self.state_interpreter = state_interpreter
+        self.state_interpret_kwargs = state_interpret_kwargs
+        self.action_interpret_kwargs = action_interpret_kwargs
+
+    def generate_order_list(self, execute_state):
+        super(RLStrategy, self).step()
+        _interpret_state = self.state_interpretor.interpret(
+            execute_result=execute_state, **self.action_interpret_kwargs
+        )
+        _policy_action = self.policy.step(_interpret_state)
```
Comment (ultmaster) [inline qlib/strategy/base.py:139]: `self.policy(_interpret_state)`

File: qlib/contrib/backtest/exchange.py [hunk 6]
```diff
@@ -8,16 +8,20 @@
...
+from ...utils.sample import sample_feature
 from ...log import get_module_logger
+from .order import Order
 
 
 class Exchange:
     def __init__(
         self,
-        trade_dates=None,
+        freq="day",
```
Comment (you-n-g) [inline qlib/contrib/backtest/exchange.py:22]: turnover limit threshing

File: qlib/utils/sample.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+                return 13 + (minute_index - 120) // 60, (minute_index - 120) % 60
+            else:
+                raise ValueError("calendar minute_index error")
+
+        if freq_raw != "minute":
+            raise ValueError("when sampling minute calendar, freq of raw calendar must be minute or min")
+        else:
+            if raw_count > sam_count:
+                raise ValueError("raw freq must be higher than sampling freq")
+        _calendar_minute = np.unique(
```
[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/utils/sample.py:101]: add docstring

File: qlib/utils/sample.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+        The Sampled DataFrame/Series/Value
+    """
+
+    selector_datetime = slice(start_time, end_time)
+    if fields is None:
+        fields = slice(None)
+
+    from ..data.dataset.utils import get_level_index
+
+    datetime_level = get_level_index(feature, level="datetime") == 0
```
[the remark this answers is not in the document]
Comment (you-n-g) [reply qlib/utils/sample.py:276]: lasy_sort_index
qlib.utils

index.is_lexsorted()

File: qlib/contrib/strategy/model_strategy.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+        if "trade_exchange" in common_infra:
+            self.trade_exchange = common_infra.get("trade_exchange")
+
+    def get_risk_degree(self, trade_index=None):
+        """get_risk_degree
+        Return the proportion of your total value you will used in investment.
+        Dynamically risk_degree will result in Market timing.
+        """
+        # It will use 95% amoutn of your total value by default
```
Comment (ultmaster) [inline qlib/contrib/strategy/model_strategy.py:81]: amount

File: qlib/strategy/base.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        if level_infra is not None:
+            self.reset_level_infra(level_infra)
+
+        if common_infra is not None:
+            self.reset_common_infra(common_infra)
+
+        if rely_trade_decision is not None:
+            self.rely_trade_decision = rely_trade_decision
+
+    def generate_trade_decision(self, execute_state):
```
Comment (ultmaster) [inline qlib/strategy/base.py:74]: What is the format/interface of execute state? What is expected to get from execute_state when I write a new strategy.

What is the interface of return value of `generate_trade_decision`?

File: qlib/strategy/base.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    def reset_common_infra(self, common_infra):
+        if not hasattr(self, "common_infra"):
+            self.common_infra = common_infra
+        else:
+            self.common_infra.update(common_infra)
+
+        if "trade_account" in common_infra:
+            self.trade_position = common_infra.get("trade_account").current
+
+    def reset(self, level_infra: dict = None, common_infra: dict = None, rely_trade_decision=None, **kwargs):
```
Comment (ultmaster) [inline qlib/strategy/base.py:59]: What is the interface of:

- level_infra
- common_infra
- rely_trade_decision

When is reset expected to be called?

**Comment by bxdd** [comment]
This pr is closed, see #438

File: qlib/utils/sample.py [hunk 12]
```diff
@@ -35,7 +35,7 @@ def parse_freq(freq: str) -> Tuple[int, str]:
         raise ValueError(
             "freq format is not supported, the freq should be like (n)month/mon, (n)week/w, (n)day/d, (n)minute/min"
         )
-    _count = int(match_obj.group(1) if match_obj.group(1) else "1")
+    _count = int(match_obj.group(1)) if match_obj.group(1) is None else 1
```
Comment (bxdd) [inline qlib/utils/sample.py]: It should be `1 if match_obj.group(1) is None else int(match_obj.group(1))`?

**Pull request closed without merging**